### ISSUE TITLE
[POLL][RFC] DataParallel Deprecation

### ISSUE THREAD

**Issue by cbalioglu** [pytorch/pytorch#65936]
Hey everyone,

We wanted to let you know that we are considering the deprecation of DataParallel (`torch.nn.DataParallel`, a.k.a. DP) module with the upcoming v1.11 release of PyTorch. Our plan is to keep DataParallel in maintenance mode for the 12 months following the v1.11 release and afterwards completely remove it from our code base. No matter whether we follow this plan or not, we still highly encourage everyone relying on DataParallel to onboard with the Distributed Data Parallel (`torch.distributed`, a.k.a. DDP) module as we consider it the future of PyTorch Distributed. Your feedback is very important to us, so please let us know your thoughts.

--

**Why should I use DDP over DP?**
DDP is a faster and scalable alternative to DP. It is in active development, has a clear roadmap, and represents the future of PyTorch Distributed. DP was our first attempt to data parallel training, but over time due to its inefficiency and lack of some key features such as multi-node training fell into disuse. Today a big majority of our users have already migrated to DDP. We ask our remaining users to migrate in the near future so that we can dedicate all our resources to DDP.

**I use DP on a single multi-GPU machine. Why do I to have to use a “distributed” API?**
DDP can be used on a single machine as well. In fact it is faster than DP on a single machine with multiple GPUs since it avoids Python’s GIL.

**How is DDP faster than DP?**
DP uses multi-threaded parallelism, while DDP uses multi-process parallelism. Due to Python’s infamous GIL contention across threads, the collective operations of DP introduce a lot of overhead and are universally slower than their DDP counterparts. Moreover DP repeatedly copies model parameters between the threads which introduces yet another overhead.

**But DP is a lot easier to use, DDP requires me to write additional code to set it up. Is there a way to avoid it?**
We are actively working on filling any usability gaps between the two APIs and aim to have parity in the following months before the removal of DP. Also note that DDP comes with a launcher script (see [here](https://pytorch.org/docs/stable/elastic/quickstart.html)) that handles most of the boilerplate code. In the meantime if you have any suggestions, feel free to reach out to us.

--

If you have any specific questions, please let us know below. Going forward we will actively monitor this post and try to address any concerns, suggestions, or feedback you have.

Thank you,
PyTorch Distributed Team

cc @mrshenli @pritamdamania @rohan-varma @vitaly-fedyunin @ngimel @bowangbj

**Comment by ananthsub** [comment]
> Also note that DDP comes with a launcher script (see here) that handles most of the boilerplate code.

@cbalioglu another popular use case for DataParallel is within Jupyter notebooks (based on PyTorch Lightning users). Is there a notebook launcher that could be supported as well? https://discuss.pytorch.org/t/distributeddataparallel-on-terminal-vs-jupyter-notebook/101404

**Comment by h6197627** [comment]
Another question what is the real state with #35191?
Is it not relevant anymore and warning is obsolete?

**Comment by vadimkantorov** [comment]
One other usecase for DP versus DDP is interactive debugging by setting `breakpoint()` in model code

**Comment by ngimel** [comment]
Thank you for your feedback @tstandley. I see a few actionable items here
1) Improve the documentation for DistributedSampler, possibly enable some warnings when set_epoch or random seed appears to be set incorrectly cc @ejguan 
2) Gracefully handle zombie processes from Ctrl-C-int ddp cc @mrshenli, @cbalioglu 
3) Guidelines and utilities for correctly doing multi-gpu validation
4) I'm not sure nccl hangs are actually actionable, but cc @mrshenli just in case
As for using extra CPU core per GPU, DP spawns extra threads, so in terms of CPU usage it should not amount to a lot of difference

**Comment by vadimkantorov** [comment]
@ngimel Here's a recent example of CPU thread oversubscription: https://github.com/pytorch/pytorch/issues/66249#issuecomment-939207348

**Comment by vadimkantorov** [comment]
@ngimel My related issue on guidance for implementing dataset evaluators: https://github.com/pytorch/pytorch/issues/64905

This is especially an issue because most of (often legacy) code for existing datasets was often written without configurable multi-process parallelization in mind

**Comment by vadimkantorov** [comment]
E.g. popular Deformable-DETR distributed scripts lead to dangling replicas and defunct processes if there are some exceptions at the beginning of training.

With DDP it's then super-hard to kill all remaining replicas without touching some other running pythons

https://github.com/fundamentalvision/Deformable-DETR/blob/main/tools/run_dist_launch.sh

**Comment by aluo-x** [comment]
There are still two minor complications for DDP that I encounter:
1. There is no official support for higher order gradients - important for GAN training with gradient penalty. See #63929
2. Sometimes DDP processes are not cleanly shutdown with Ctrl + C, which requires `pkill -f "filename.py"` - also noted in two comments above

I'm still in favor of maintaining DataParallel, for backwards compatibility with the huge amount of existing pytorch code if nothing else.

**Comment by vadimkantorov** [comment]
It would be nice to have random port selection by default: https://github.com/pytorch/pytorch/issues/73320, without having to set up num_proc_per_node beforehand and without passing rdv-related changes (as proposed in https://github.com/pytorch/pytorch/pull/73598)

**Comment by micahcc** [comment]
FWIW Today I was pained by debugging with DDP and turned my trainer back to DataParallel and got a 25% training time speedup. For very straightforward neural networks I don't know that DDP makes sense. I can't share code because its internal, but if you take the torchvision detector trainers and test them I suspect you would find something similar.

**Comment by ex3ndr** [comment]
In my learning experience with two RTX 4090 without P2P support DP is much better and easier to use than DDP, which is really beneficial only when you have P2P. Also as a casual trainer DP is much easier to setup, unlike DDP.

**Comment by Puiching-Memory** [comment]
Does the removal of GIL from Python 3.13t's experimental features give us hope of reconsidering DP mode?
